// tslint-eslint-rules/dist/rules/sortImportsRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint");
var ts = require("typescript");
var RULE_NAME = 'sort-imports';
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var walker = new RuleWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    };
    return Rule;
}(Lint.Rules.AbstractRule));
Rule.metadata = {
    ruleName: RULE_NAME,
    description: 'enforce sorting import declarations within module',
    rationale: (_a = ["\n      When declaring multiple imports, a sorted list of import declarations make it easier for developers to \n      read the code and find necessary imports later. This rule is purely a matter of style.\n      \n      This rule checks all import declarations and verifies that all imports are first sorted by the used member \n      syntax and then alphabetically by the first member or alias name.\n      "], _a.raw = ["\n      When declaring multiple imports, a sorted list of import declarations make it easier for developers to \n      read the code and find necessary imports later. This rule is purely a matter of style.\n      \n      This rule checks all import declarations and verifies that all imports are first sorted by the used member \n      syntax and then alphabetically by the first member or alias name.\n      "], Lint.Utils.dedent(_a)),
    optionsDescription: (_b = ["\n      - `\"ignore-case\"` does case-insensitive comparisons (default: `false`)\n      - `\"ignore-member-sort\"` allows members in multiple type imports to occur in any order (default: `false`)\n      - `\"member-syntax-sort-order\"` (default: `[\"none\", \"all\", \"multiple\", \"single\", \"alias\"]`); all 5 items must be \n      present in the array, but you can change the order: \n        - `none` = import module without exported bindings.\n        - `all` = import all members provided by exported bindings.\n        - `multiple` = import multiple members.\n        - `single` = import a single member.\n        - `alias` = creates an alias for a member. This is unique to TER and not in ESLint's `sort-imports`.\n      "], _b.raw = ["\n      - \\`\"ignore-case\"\\` does case-insensitive comparisons (default: \\`false\\`)\n      - \\`\"ignore-member-sort\"\\` allows members in multiple type imports to occur in any order (default: \\`false\\`)\n      - \\`\"member-syntax-sort-order\"\\` (default: \\`[\"none\", \"all\", \"multiple\", \"single\", \"alias\"]\\`); all 5 items must be \n      present in the array, but you can change the order: \n        - \\`none\\` = import module without exported bindings.\n        - \\`all\\` = import all members provided by exported bindings.\n        - \\`multiple\\` = import multiple members.\n        - \\`single\\` = import a single member.\n        - \\`alias\\` = creates an alias for a member. This is unique to TER and not in ESLint's \\`sort-imports\\`.\n      "], Lint.Utils.dedent(_b)),
    options: {
        type: 'object',
        properties: {
            'member-syntax-sort-order': {
                type: 'array',
                items: {
                    type: 'string',
                    enum: ['none', 'all', 'multiple', 'single', 'alias']
                },
                minLength: 5,
                maxLength: 5
            },
            'ignore-case': {
                type: 'boolean'
            },
            'ignore-member-sort': {
                type: 'boolean'
            }
        }
    },
    optionExamples: [
        (_c = ["\n        \"", "\": [true]\n        "], _c.raw = ["\n        \"", "\": [true]\n        "], Lint.Utils.dedent(_c, RULE_NAME)),
        (_d = ["\n        \"", "\": [true, { \"ignore-case\" }]\n        "], _d.raw = ["\n        \"", "\": [true, { \"ignore-case\" }]\n        "], Lint.Utils.dedent(_d, RULE_NAME)),
        (_e = ["\n        \"", "\": [true, { \"ignore-member-sort\" }]\n        "], _e.raw = ["\n        \"", "\": [true, { \"ignore-member-sort\" }]\n        "], Lint.Utils.dedent(_e, RULE_NAME)),
        (_f = ["\n        \"", "\": [true, { \"member-syntax-sort-order\": [\"all\", \"single\", \"multiple\", \"none\", \"alias\"] }]\n        "], _f.raw = ["\n        \"", "\": [true, { \"member-syntax-sort-order\": [\"all\", \"single\", \"multiple\", \"none\", \"alias\"] }]\n        "], Lint.Utils.dedent(_f, RULE_NAME))
    ],
    typescriptOnly: false,
    type: 'style'
};
exports.Rule = Rule;
var MemberSyntaxType;
(function (MemberSyntaxType) {
    MemberSyntaxType[MemberSyntaxType["None"] = 0] = "None";
    MemberSyntaxType[MemberSyntaxType["All"] = 1] = "All";
    MemberSyntaxType[MemberSyntaxType["Multiple"] = 2] = "Multiple";
    MemberSyntaxType[MemberSyntaxType["Single"] = 3] = "Single";
    MemberSyntaxType[MemberSyntaxType["Alias"] = 4] = "Alias";
})(MemberSyntaxType || (MemberSyntaxType = {}));
var RuleWalker = (function (_super) {
    __extends(RuleWalker, _super);
    function RuleWalker(sourceFile, options) {
        var _this = _super.call(this, sourceFile, options) || this;
        _this.currentImportIndex = 0;
        var optionSet = _this.getOptions()[0] || {};
        _this.ignoreCase = _this.hasOption('ignore-case');
        _this.ignoreMemberSort = _this.hasOption('ignore-member-sort');
        _this.expectedOrder = RuleWalker._processMemberSyntaxSortOrder(optionSet['member-syntax-sort-order']);
        _this.currentSortValue = { sortValue: '', originalValue: '' };
        if (_this.ignoreCase) {
            _this.caseConverter = function (s) { return s.toUpperCase(); };
        }
        else {
            _this.caseConverter = function (s) { return s; };
        }
        return _this;
    }
    RuleWalker.prototype.visitNode = function (node) {
        if (node.kind === ts.SyntaxKind.ImportDeclaration ||
            node.kind === ts.SyntaxKind.ImportEqualsDeclaration) {
            this._validateOrder(node);
        }
        _super.prototype.visitNode.call(this, node);
    };
    RuleWalker.prototype.visitNamedImports = function (node) {
        if (!this.ignoreMemberSort) {
            this._validateMemberSort(node);
        }
        _super.prototype.visitNamedImports.call(this, node);
    };
    RuleWalker.prototype._validateMemberSort = function (node) {
        var _this = this;
        var imports = node.elements.map(function (e) { return _this.caseConverter(e.getText()); });
        var importReduction = imports.reduce(function (prev, current) { return prev + current; });
        var sortedImports = imports.sort();
        var sortedReduction = sortedImports.reduce(function (prev, current) { return prev + current; });
        if (importReduction !== sortedReduction) {
            this.addFailureAtNode(node, 'Member imports must be sorted alphabetically.');
        }
    };
    RuleWalker.prototype._validateOrder = function (node) {
        var importData = this._determineImportType(node);
        if (importData) {
            var index = this.expectedOrder.indexOf(importData.memberSyntaxType, this.currentImportIndex);
            if (index !== -1) {
                if (this.expectedOrder[this.currentImportIndex] !== importData.memberSyntaxType) {
                    this.currentImportIndex = index;
                    this.currentSortValue = {
                        sortValue: this.caseConverter(importData.sortValue),
                        originalValue: importData.sortValue
                    };
                }
                else if (this.currentSortValue.sortValue > this.caseConverter(importData.sortValue)) {
                    this.addFailureAtNode(node, "All imports of the same type must be sorted alphabetically. \"" + importData.sortValue + "\" must come before \"" + this.currentSortValue.originalValue + "\"");
                }
                else {
                    this.currentSortValue = {
                        sortValue: this.caseConverter(importData.sortValue),
                        originalValue: importData.sortValue
                    };
                }
            }
            else {
                var currentSyntaxType = MemberSyntaxType[importData.memberSyntaxType];
                var previousSyntaxType = MemberSyntaxType[this.expectedOrder[this.currentImportIndex]];
                this.addFailureAtNode(node, "All imports of type \"" + currentSyntaxType + "\" must occur before all imports of type \"" + previousSyntaxType + "\"");
            }
        }
        else {
            this.addFailureAtNode(node, 'Could not determine import type');
        }
    };
    RuleWalker.prototype._determineImportType = function (node) {
        var nodeText = node.getFullText();
        if (node.kind === ts.SyntaxKind.ImportEqualsDeclaration) {
            var aliasMatch = /\bimport\s+(\w+)\s*=.+/g.exec(nodeText);
            return {
                memberSyntaxType: MemberSyntaxType.Alias,
                sortValue: aliasMatch[1]
            };
        }
        else {
            var singleMatch = /\bimport\s+(?:{?([^,{}\*]+?)}?)\s*from\s+[\'"](?:[^"\']+)["\']/g.exec(nodeText);
            var multipleMatch = /\bimport\s*{?\s*([^{}\'",]+?)\s*,(?:\s*.+\s*,\s*)*\s*.+\s*}?\s*from\s+[\'"](?:[^"\']+)["\']/g.exec(nodeText);
            var noneMatch = /\bimport\s+[\'"]([^"\']+)["\']/g.exec(nodeText);
            var allMatch = /\bimport\s+\*\s+as\s+(.+)\s+from\s+[\'"](?:[^"\']+)["\']/g.exec(nodeText);
            var result = void 0;
            if (singleMatch !== null) {
                result = {
                    memberSyntaxType: MemberSyntaxType.Single,
                    sortValue: singleMatch[1]
                };
            }
            else if (multipleMatch !== null) {
                result = {
                    memberSyntaxType: MemberSyntaxType.Multiple,
                    sortValue: multipleMatch[1]
                };
            }
            else if (noneMatch !== null) {
                result = {
                    memberSyntaxType: MemberSyntaxType.None,
                    sortValue: noneMatch[1]
                };
            }
            else if (allMatch !== null) {
                result = {
                    memberSyntaxType: MemberSyntaxType.All,
                    sortValue: allMatch[1]
                };
            }
            return result;
        }
    };
    RuleWalker._processMemberSyntaxSortOrder = function (sortOption) {
        var defaultOrder = [MemberSyntaxType.None, MemberSyntaxType.All, MemberSyntaxType.Multiple, MemberSyntaxType.Single, MemberSyntaxType.Alias];
        if (Array.isArray(sortOption) && typeof sortOption[0] === 'string' && sortOption.length === 5) {
            var memberSyntaxTypeMap_1 = {
                none: MemberSyntaxType.None,
                all: MemberSyntaxType.All,
                multiple: MemberSyntaxType.Multiple,
                single: MemberSyntaxType.Single,
                alias: MemberSyntaxType.Alias
            };
            var order_1 = [];
            var usedOptions_1 = {};
            sortOption.forEach(function (t) {
                if (usedOptions_1[t] !== undefined) {
                }
                else {
                    usedOptions_1[t] = t;
                    if (memberSyntaxTypeMap_1[t]) {
                        order_1.push(memberSyntaxTypeMap_1[t]);
                    }
                }
            });
            return order_1;
        }
        else {
            return defaultOrder;
        }
    };
    return RuleWalker;
}(Lint.RuleWalker));
var _a, _b, _c, _d, _e, _f;
